dmoj/judge.py

```diff
@@ -174,108 +93,130 @@ def begin_grading(
-                yield BatchBegin()
+    def _ipc_grading_begin(self, _report, is_pretested):
+        self.packet_manager.begin_grading_packet(is_pretested)
 
-                for batched_case in self.grade_cases(
-                    grader,
-                    case.batched_cases,
-                    short_circuit=case.config['short_circuit'],
-                    is_short_circuiting=is_short_circuiting,
-                ):
-                    # A batched case just failed.
-                    # There are two cases where this means that we should completely short-circuit:
-                    # 1. If the batch was worth 0 points, to emulate the property of 0-point cases.
-                    # 2. If the short_circuit flag is true, see <https://github.com/DMOJ/judge/issues/341>.
-                    if (batched_case.result_flag & Result.WA) and (not case.points or short_circuit):
-                        is_short_circuiting = True
-                    yield batched_case
-                yield BatchEnd()
-                continue
-
-            # Stop grading if we're short circuiting
-            if is_short_circuiting:
-                result = Result(case)
-                result.result_flag = Result.SC
-                yield result
-                continue
-
-            result = grader.grade(case)
-
-            # If the submission was killed due to an user-initiated abort, any result is meaningless.
-            if self._terminate_grading:
-                raise TerminateGrading()
-
-            # If the WA bit of result_flag is set and we are set to short-circuit (e.g., in a batch),
-            # short circuit the rest of the cases.
-            # Do the same if the case is a pretest (i.e. has 0 points)
-            if (result.result_flag & Result.WA) > 0 and (short_circuit or not case.points):
-                is_short_circuiting = True
-
-            yield result
-
-    def log_internal_error(self, exc=None):
-        # If exc is exists, raise it so that sys.exc_info() is populated with its data
-        if exc:
-            try:
-                raise exc
-            except:  # noqa: E722, we want to catch everything
-                pass
-        exc = sys.exc_info()
+    def _ipc_grading_end(self, _report):
+        self.packet_manager.grading_end_packet()
 
-        message = ''.join(traceback.format_exception(*exc))
+    def _ipc_result(self, report, batch_number, case_number, result):
+        codes = result.readable_codes()
 
-        # Strip ANSI from the message, since this might be a checker's CompileError
-        # ...we don't want to see the raw ANSI codes from GCC/Clang on the site.
-        # We could use format_ansi and send HTML to the site, but the site doesn't presently support HTML
-        # internal error formatting.
-        self.packet_manager.internal_error_packet(strip_ansi(message))
+        is_sc = result.result_flag & Result.SC
+        colored_codes = list(
```

Comment:
Maybe we should just use a list comprehension here instead of `list(map(...))`.

Reply:
Done.